### PR TITLE
Fix latents sampling from image not working with Autoencoders that produce latents directly (not latent distribution)

### PR DESCRIPTION
**Pull request by slep0v**
# What does this PR do?

(This PR is a reopened #4810 due to mistake in branch rebase)

Fixes the issue in `prepare_latents` methods of pipelines that have an image as input (img2img/inpaint + controlnet versions) not working with Autoencoders that unlike AutoencoderKL produce latents directly (i.e. `AutoencoderTiny`).

This is done by introducing a `produce_latents` function that checks whether `AutoencoderOutput` has a ready latent vector attribute or only a distribution which needs to be sampled from.

This issue was addressed in #4720 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Great! Good for merge to me once fast tests pass :-) 

Thanks for opening a new clean PR! 

@sayakpaul can you also check?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py [hunk 1]
```diff
@@ -91,6 +91,16 @@
...
 
+# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.retrieve_latents
+def retrieve_latents(encoder_output: BaseOutput, generator):
+    if hasattr(encoder_output, 'latent_dist'):
+        return encoder_output.latent_dist.sample(generator)
+    elif hasattr(encoder_output, 'latents'):
+        return encoder_output.latents
+    else:
+        raise AttributeError('Could not access latents of provided encoder_output')
+
```
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py:102]: We use double quotes.

**Review by sayakpaul — APPROVED** [review]
Looks great!

Let's make sure to make the CI green :)

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py (hunk 1 above)
Comment (slep0v) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py]: Fixed

**Comment by sayakpaul** [comment]
@slep0v sorry for skipping this one. Would you maybe like to reopen it?

**Comment by sayakpaul** [comment]
There are some merge conflicts here and I think it's better to work in a new PR. Could you provide your commit email id so that we could add you as a co-author of the future commits related to this?

**Comment by slep0v** [comment]
Ok. `37597789+slep0v@users.noreply.github.com` - I suppose you asked for this email. 
Or do you need my primary email that is hidden by default in github?

**Comment by sayakpaul** [comment]
That email should work. Thank you!

**Comment by sayakpaul** [comment]
Closing it in the spirit of https://github.com/huggingface/diffusers/pull/5636 where @slep0v has been added as a co-author.

**Pull request closed without merging**